Improve: add door symbol display to custom exits lines and to stubs

SlySven opened the pull request:
This PR will also allow colors for open/closed/locked doors to be customised in the future (by adjusting the new `T2DMap` `QColor` members: `mOpenDoorColor`, `mClosedDoorColor` and `mLockedDoorColor`.

Moved the `QMap<int, QPoint> mAreaExitsList` member out of the class header - it is now a local `areaExitsList` value held in the `paintEvent(...)` method and passed by reference to a couple of paint helper methods.

Revise the code that painted exit stubs as the formulas being used to work out the inner and outer points on the line were wrong - and as they were dependent on the room size the stubs changed as the room size did. Since we may now be painting a door marker on them it is easier if the length is constant and does not depend on the room size.

Since we are now able to show door markings on custom exit lines (including on special exits) and stub exits the tool-tips in the room exits control panel are updated to now reflect this.

This will close #499 which is the same as an issue that **Fuligin** raised in the Discord **#help** channel on 2021/01/08 and close #668.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: add door symbol display to custom exits lines and to stubs

This PR will also allow colors for open/closed/locked doors to be
customised in the future (by adjusting the new `T2DMap` `QColor`
members: `mOpenDoorColor`, `mClosedDoorColor` and `mLockedDoorColor`.

Moved the `QMap<int, QPoint> mAreaExitsList` member out of the class header
- it is now a local `areaExitsList` value held in the `paintEvent(...)`
method and passed by reference to a couple of paint helper methods.

Revise the code that painted exit stubs as the formulas being used to work
out the inner and outer points on the line were wrong - and as they were
dependent on the room size the stubs changed as the room size did. Since
we may now be painting a door marker on them it is easier if the length is
constant and does not depend on the room size.

Since we are now able to show door markings on custom exit lines (including
on special exits) and stub exits the tool-tips in the room exits control
panel is updated to now reflect this.

This will close #499 and close #668 and address an issue that Fuligin
raised in the Discord # help channel on 2021/01/08 .

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Also a comment I made in another issue https://github.com/Mudlet/Mudlet/issues/2427#issuecomment-478082367 describes the sort of thing that I have *finally* done...

SlySven committed:
BugFix: revise handling of area exits with doors

This actually doubles the amount that area exits stick out from the
starting room and revises the drawing code so that it is not so dependent
on the room size (which modified the `exitWidth` local that the code was
using). The particular detail that required this commit was that the way
that the underlying line was drawn that was used to derive the door symbol
from had the start and end points the opposite way around to that which I
had developed the code for!

Whilst cleaning things up I spotted that there was a redundant conversion
from a `QPointF` to a `QPoint` and then back to a `QPointF` when handling
the area exit click location - that is used to record when an area exit is
(double) clicked upon to initiate a speedwalk to the room in the adjacent
area.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Delwing reviewed (approved):
Looks good imho.

Although I must admit that I would like non-curve doors.

vadi2 commented:
What would you like to see?

Delwing commented:
Some straight lines :) But this is just a matter of taste I guess.

SlySven commented:
They were shaped so that when the two rooms are close together and only one direction has a door it is clear to see which direction it applies to...

vadi2 commented:
That's a good point, but majority of the doors are two way, and it does look like a funny little chromosome now. Do you think it can be improved?

Delwing commented:
How about shape like that? Line with a little triangle in the middle indicating way in case of one way door.

```
|
|\
|/
|
```

SlySven commented:
Actually the simplest modification would be to take out the middle section so that you have a pair of "outward opening" doors, either side of the exit line:
```
 \
===== <- Exit line (this side is away from exit room
 /
```

> ... Line with a little triangle in the middle indicating way in case of one way door.

Since the two directions are handled separately (as they are each associated with the room the exit leaves from) trying to do that would get more complicated and involved - it would require trying to determine whether the exit was an area exit (which by definition is treated as one way even when it isn't) and whether it was a two or one way exit otherwise...

SlySven committed:
Revise: provide "pair of doors" representation for doors on 2D Map

As requested during review.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Delwing commented:
Something more happened to map rendering code.
Fresh PTB on the left (same on release version)
This branch's build on the right.
Looks like some out of view decisions come too early additionally this broken line to that location... they persist even on the middle and when moving room.

![image](https://user-images.githubusercontent.com/3740628/105226566-6ede9100-5b60-11eb-8e9b-9fb6a3ee4ead.png)
![image](https://user-images.githubusercontent.com/3740628/105226702-a5b4a700-5b60-11eb-9ef6-41b11b423ec7.png)
![image](https://user-images.githubusercontent.com/3740628/105226792-bebd5800-5b60-11eb-87d7-4378e7ad43ef.png)

SlySven commented:
I agree something is wrong - note that when the destination room is off the map then there should now still be *half a line* going towards it from the room that is on the map - in the past I think there could have been some cases where the whole line was missed out.

:thinking: 

... also, it does seem to be a bit arbitrary as to the room(s) that this is going wrong for!

Delwing commented:
Yeah, strange. 
Whole line gets missing sometimes for custom lines (usually when target room is out of view if I recall correctly), but for regular exits haven't had any issues recently.

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

SlySven commented:
@Delwing do you have a small test map that reproduces this. I've gone looking for it and haven't spotted the bug currently but I agree it is there somewhere...

vadi2 commented:
You can find it in https://github.com/IRE-Mudlet-Mapping/AchaeaCrowdmap/blob/gh-pages/Map/map, have a look at room 11122.

Delwing commented:
Sure, 
http://arkadia.kamerdyner.net/master3/map_master3.dat
Room 5172.

SlySven commented:
For the rooms where there are NO exit lines showing check for custom exit lines ***with no points***! I've just found a room like that in my map:
![image](https://user-images.githubusercontent.com/6163092/106327577-464c4a80-6276-11eb-98dd-05a6ddb3a0b7.png)
after removing that useless line:
![image](https://user-images.githubusercontent.com/6163092/106327693-7b589d00-6276-11eb-99fb-e3224c424984.png)

SlySven commented:
Ah, yours is slightly different @Delwing - it does have a valid custom line:
![image](https://user-images.githubusercontent.com/6163092/106328408-aee7f700-6277-11eb-9e2b-7d547e1f2ca1.png)
but if you zoom in too much so that the far end of the custom line is off-screen, it makes ALL the exits disappear:
![image](https://user-images.githubusercontent.com/6163092/106328534-e656a380-6277-11eb-818d-1c9538ad9376.png)

SlySven commented:
The missing exits bug has been fixed separately in #4716 .

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

Needed to pull in exit line drawing bug fix.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

Needed to include fix for crashing on exit in TTimerUnit destructor.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: tweak some aspects of the door markings

Make them a bit (150%) thicker and a little longer.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Jieiku commented:
This looks exciting!

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

Conflicts resolved in:
* src/T2DMap.h
* src/TMap.cpp
* src/TRoom.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung reviewed (approved):
The UI text changes seem good to me. I didn't delve into the C++ part, though.

SlySven committed:
Merge branch 'development' into Enhance_addDoorSymbolDisplayToCustomExitsLinesAndToStubs

SlySven committed:
Merge: pull in changes in `development` branch

Conflicts resolved in:
* src/T2DMap.cpp
* src/T2DMap.h
* src/TRoom.cpp
* src/TRoom.h
* src/dlgRoomExits.cpp
* src/ui/room_exits.ui

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: two spelling issues in comments

In fact one of the terms "statics" was correct because it was referring to
multiple variables of a certain scope, but rephrasing the sentence avoids
offending the *smelling misteak* Bot!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: use QLineF/QPointF instead of QLine/QPoint to pacify clang-tidy Bot

It seemed to be unhappy about "narrowing conversions" of `double`s to
`int`s. when a local line and point with integer coordinates are created
- and which are then used to create others which have floating point
values.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
I say merge it and see how people take to this!

SlySven commented:
Okay, I'll publish and be damned...!

Jieiku commented:
Very cool! do these changes perhaps prelude to: https://github.com/Mudlet/Mudlet/issues/2427
I remember you saying that you had some code you were working on that would allow for this, but you had not yet finished it.

SlySven commented:
> I remember you saying that you had some code you were working on that would allow for this, but you had not yet finished it.

I think that was a long time ago on a PC ~~far, far~~ *no longer in use*! Even if I could get into a spot of archaeology I suspect I'd be better off starting from scratch with the 2D map code as it is now - OTOH it is nearly Christmas and I do have more spare time at the moment so maybe I get around to it.
